### ISSUE TITLE
Implication of Recommended/Not Recommended

### ISSUE THREAD

**Issue by ekr** [tlswg/tls13-spec#1214]
There has been a bunch of discussion about what we should do with the Recommended Column

See, especially:
https://mailarchive.ietf.org/arch/msg/tls/_1fPsfy6kkQwFAtX3eKeIkRKK4E/

This is a placeholder issue to workshop an approach.

**Comment by richsalz** [comment]
The initial impetus for 'recommended' was to allow national ciphers to register in IANA, but to not give them the imprimateur of officially-recommended status. That's still important and we shouldn't lose that.  One possibility is to say N unless it's standards-track and any s-t document must explicitly say what value it wants.

**Comment by ekr** [comment]
@chris-wood

**Comment by ekr** [comment]
My sense is that part of the impetus here now is to be able to express that certain algorithms that are disrecommended for general use (e.g., integrity only)

**Comment by davidben** [comment]
In addition to whether the WG "officially" recommends them, we probably need to say something about what properties TLS claims to provide, and what assumptions other documents can make. Things like:

* Appendix E talks about protecting endpoint identities. Obviously that is only true in so far as the AEAD isn't secretly broken, but if the non-recommended cipher suite doesn't even try, I don't think that's Appendix E's problem.
* Appendix C.4's tracking protections depend on confidentiality. Since confidentiality is part of RFC8446's guarantees, C.4 shouldn't have to qualify that with "assuming you are not in a non-recommended mode that skips confidentiality". Same with ECH.
* Say some extension document says "compute ABC by passing XYZ into the chosen cipher suite's AEAD". If some non-recommended cipher suite isn't even based on AEADs, the document shouldn't need to qualify the instructions.

If we allow registrations for these sorts of things, we probably want a way to say the code points are suitably out of scope. (No opinions on whether that's recommended=N, or some new third state.)

**Comment by ekr** [comment]
I would suggest that we have four categories (names TBD)

- Recommended: The WG thinks this is good for general use.
  (E.g., TLS_AES_128_GCM_SHA256). Requires a standards track
  RFC.
  
- Conditionally Recommended: The WG thinks this is good for limited
  scope, as with IoT (must come with some description of that scope)
  (e.g., TLS_AES_128_CCM_8_SHA256). Requires a standards
  track RFC.
  
- Not recommended: The WG expresses no opinion on this
  (e.g., TLS_GOSTR341112_256_WITH_KUZNYECHIK_CTR_OMAC)
  
- Discouraged (?): Known not to provide the rated TLS 1.3 security
  guarantees (e.g., TLS_SHA256_SHA256)

What do people think about this? I do worry about some slop between
"conditionally recommended" and "discouraged" but I think the
experts/WG can sort it out.

**Comment by davidben** [comment]
Works for me. (Mechanically, this means that code point allocations still need to go through some cursory TLSWG consult to get a category?)

**Comment by richsalz** [comment]
Where do you put integrity-only ciphers?

Maybe discouraged is discouraged/obsolete?

**Comment by ekr** [comment]
I would put integrity-only in "Discouraged"

**Comment by martinthomson** [comment]
How would a suite end up in that state? Known broken cipher? Integrity-only? What else?

**Comment by ekr** [comment]
Maybe known weak, e.g., a curve that had less than ~128 bits of security.

**Comment by kaduk** [comment]
> Say some extension document says "compute ABC by passing XYZ into the chosen cipher suite's AEAD". If some non-recommended cipher suite isn't even based on AEADs, the document shouldn't need to qualify the instructions.

(We almost ended up with something like that for DTLS record protection schemes, but took a different route.)

In Ekr's proposal from https://github.com/tlswg/tls13-spec/issues/1214#issuecomment-782373964 there are four categories, two of which require standards-track RFCs and thus are fairly straightforward to figure out how to get classified into, procedurally wise.  For the other two categories, davidben rightly notes that some classification procedure is needed.  The most natural procedure is "just let the DEs decide, but that lacks the imprimateur of the WG or the IETF as a whole.  One could perhaps say that the DEs have to consult the WG list, but it gets pretty awkward if we don't leave the decision in the hands of the DEs (which would include, say, having the WG chairs judge WG consensus).  I'm pretty wary of anything that puts more IETF process in the way of external registrations than the current specification-required review, due to the risk of being too high a bar to get used in practice.

**Comment by ekr** [comment]
On Mon, Feb 22, 2021 at 10:24 PM kaduk <notifications@github.com> wrote:

> Say some extension document says "compute ABC by passing XYZ into the
> chosen cipher suite's AEAD". If some non-recommended cipher suite isn't
> even based on AEADs, the document shouldn't need to qualify the
> instructions.
>
> (We almost ended up with something like that for DTLS record protection
> schemes, but took a different route.)
>
> In Ekr's proposal from #1214 (comment)
> <https://github.com/tlswg/tls13-spec/issues/1214#issuecomment-782373964>
> there are four categories, two of which require standards-track RFCs and
> thus are fairly straightforward to figure out how to get classified into,
> procedurally wise. For the other two categories, davidben rightly notes
> that some classification procedure is needed. The most natural procedure is
> "just let the DEs decide, but that lacks the imprimateur of the WG or the
> IETF as a whole. One could perhaps say that the DEs have to consult the WG
> list, but it gets pretty awkward if we don't leave the decision in the
> hands of the DEs (which would include, say, having the WG chairs judge WG
> consensus). I'm pretty wary of anything that puts more IETF process in the
> way of external registrations than the current specification-required
> review, due to the risk of being too high a bar to get used in practice.
>
Agreed. However, note that it's possible to reclassify something, too. I
would suggest that we leave it in the hands of the DEs with the potential
of appeal to the WG if you don't like that answer. You'd have your
registration in any case, so these could proceed in parallel.

-Ekr

**Comment by chris-wood** [comment]
@seanturner confirming your comment during IETF 110: this is a problem for RFC8447, right?

**Comment by dkg** [comment]
Recommended/Not Recommended/Discouraged map neatly into implementation's client or server API:  recommended is enabled by default, not recommended can be enabled explicitly by the user of the API; and discouraged cannot be enabled at all.

"conditionally recommended" is problematic from an API perspective though, because the range of conditions could be so vast.

It's easy to imagine an implementer just lumping "conditionally recommended" in with "recommended", which would be a bad outcome.

Do we actually need the "conditionally recommended" status?  or can we just go with recommended/not recommended/discouraged?

**Comment by richsalz** [comment]
I like your mapping, @dkg.  I think "conditionally" is ifdef in the source.

**Comment by dkg** [comment]
@richsalz `ifdef` in the source sounds good to me for "conditionally recommended" -- but it's not sufficient.  if the `#define` is set, should the algorithm in question be enabled by default?  or should a runtime switch still be required?

if the answer is that a runtime switch is still required, that puts "conditionally recommended" in some sense *below* "not recommended" -- because it needs to be explicitly enabled both at build and runtime, whereas not recommended might only need the runtime switch.  OTOH, if the answer is that no runtime switch is required, then you're going to have distributed binary TLS stacks that automatically enable things that might be inappropriate for their application.

**Comment by richsalz** [comment]
I'm thinking like an OpenSSL developer, you're thinking like an application developer.  I think.

**Comment by ekr** [comment]
This is going to be resolved in 8447-bis, I think, so I propose to close this,

**Comment by ekr** [comment]
Discussed in Vienna. Closed.
